common4j/src/main/com/microsoft/identity/common/java/challengehandlers/PKeyAuthChallenge.java

```diff
@@ -45,163 +45,144 @@
-        private String mNonce = "";
-        private String mContext = "";
-        private List<String> mCertAuthorities;
-        private String mThumbprint = "";
-        private String mVersion;
-        private String mSubmitUrl;
-
-        public Builder setNonce(final String nonce) {
-            mNonce = nonce;
-            return self();
-        }
-
-        public Builder setContext(final String context) {
-            mContext = context;
-            return self();
-        }
-
-        public Builder setCertAuthorities(final List<String> certAuthorities) {
-            mCertAuthorities = certAuthorities;
-            return self();
-        }
-
-        public Builder setThumbprint(final String thumbprint) {
-            mThumbprint = thumbprint;
-            return self();
-        }
+    public Map<String, String> getChallengeHeader() throws ClientException {
+        final String methodName = ":getChallengeHeader";
 
-        public Builder setVersion(final String version) {
-            mVersion = version;
-            return self();
+        // Either cert thumbprint or authorities must present
+        // Otherwise, we won't have enough information to look for the cert.
+        //
+        // In such case, Client should ideally ignore this scenario /
+        // send a response which is equivalent to no certificate present on client.
+        if ((mCertAuthorities == null || mCertAuthorities.size() == 0) &&
+                StringUtil.isNullOrEmpty(mThumbprint)) {
+            Logger.info(TAG + methodName,
+                    "Both cert Authorities and Thumbprint are not provided." +
+                            "Sending a response which is equivalent to no certificate present on client.");
+            return getChallengeHeaderWithoutSignedJwt();
         }
 
-        public Builder setSubmitUrl(final String submitUrl) {
-            mSubmitUrl = submitUrl;
-            return self();
-        }
+        // If not device cert exists, alias or private key will not exist on the device
+        // Suppressing unchecked warnings due to the generic type not provided in the object returned from method getDeviceCertificateProxy
+        @SuppressWarnings(WarningType.unchecked_warning)
+        final Class<IDeviceCertificate> certClazz =
+                (Class<IDeviceCertificate>) AuthenticationSettings.INSTANCE.getDeviceCertificateProxy();
 
-        public Builder self() {
-            return this;
+        if (certClazz == null) {
+            Logger.warn(TAG + methodName, "Device Certificate Proxy is not initialized");
+            return getChallengeHeaderWithoutSignedJwt();
         }
 
-        public PKeyAuthChallenge build() {
-            return new PKeyAuthChallenge(this);
+        final IDeviceCertificate deviceCertProxy = getWPJAPIInstance(certClazz);
+        if (!deviceCertProxy.isValidIssuer(mCertAuthorities)
+                && !StringUtil.equalsIgnoreCase(deviceCertProxy.getThumbPrint(), mThumbprint)) {
```

Comment:
should this be an or operator?

Reply:
and. if we cannot validate against both, then we do nothing.

Reply:
let me rewrite this to make it easier to read.